plugin/v1/healthz.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"context"
+	"fmt"
+
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/golang/glog"
+	kmspb "google.golang.org/api/cloudkms/v1"
+	"google.golang.org/grpc"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/GoogleCloudPlatform/k8s-cloudkms-plugin/plugin"
+)
+
+type HealthZ struct {
+	plugin.HealthZ
+}
+
+func NewHealthChecker(keyName string, keyService *kmspb.ProjectsLocationsKeyRingsCryptoKeysService,
+	unixSocketPath string, callTimeout time.Duration, servingURL *url.URL) *HealthZ {
+
+	return &HealthZ{
+		plugin.HealthZ{
+			KeyName:        keyName,
+			KeyService:     keyService,
+			UnixSocketPath: unixSocketPath,
+			CallTimeout:    callTimeout,
+			ServingURL:     servingURL,
+		},
+	}
+
+}
+
+// Serve creates http server for hosting healthz.
+func (h *HealthZ) Serve() chan error {
+	errorChan := make(chan error)
+	mux := http.NewServeMux()
+	mux.HandleFunc(fmt.Sprintf("/%s", h.ServingURL.EscapedPath()), func(w http.ResponseWriter, r *http.Request) {
+		ctx, cancel := context.WithTimeout(r.Context(), h.CallTimeout)
+		defer cancel()
+
+		connection, err := dialUnix(h.UnixSocketPath)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusServiceUnavailable)
+			return
+		}
+		defer connection.Close()
+
+		c := NewKeyManagementServiceClient(connection)
```

Comment:
Same here. This is the previous implementation v1 left in tact, I didn't try to change/improve it in this PR.

Reply:
Since there's so many changes, I'm wondering if we should forgo this v1 vs v2 and release a new version that only supports the new protocol. Customers who want the v1 behavior can use an older version of the plugin.

Reply:
It is better for project owners to decide. My 2 cents: I think that having support for 2 versions available in the same binary (temporarily) is beneficial. The V2 is not yet on by default in K8s 1.28, in 1.29 it will be on by default but V1 will still be possible and only after that the V1 support will be dropped in K8s. Having multiple binaries (built from historical revisions of branches) to juggle with to arrange a migration sounds less convenient than a flag which also lets running two containers with different KMS versions simultaneously if needed (something similar is happening in GKE, as I understand).
The V1 support would also allow running older K8s versions with latest Go runtime, which resolves many CVE vulnerabilities.
Once the customers confirm the successful adoption of V2 and when V1 is dropped in Kubernetes we'll be able to clean up the code.
And the PR is really not as invasive as it looks. The V1 files have been mostly been moved without any changes.